### ISSUE TITLE
Provide batteries-included transformer and GPT APIs with overridable components

### ISSUE THREAD

**Issue by samifouad** [dekaruntime/dsc#22]
## Goal

Make a small transformer language model expressible in clear Deka while providing complete default implementations. A user should be able to train a basic GPT without defining tensors, attention, autodiff, optimizers, batching, or sampling from scratch.

The target is a small causal language model that can train on a text file and generate continuations. Chat formatting and instruction tuning are later layers; a next-token model is not automatically a chat model.

## Default API direction

\`\`\`deka
import ai

let tokenizer = ai.CharTokenizer.from_text(training_text)?
let model = ai.GPT.create(ai.GPTConfig {
  vocabulary_size: tokenizer.size()
  context_length: 256
  embedding_size: 384
  head_count: 6
  layer_count: 6
})

let trainer = ai.Trainer.default(model, dataset)
trainer.train(steps: 5000)?
let text = model.generate("ROMEO:", tokens: 100)?
\`\`\`

The default implementation owns token and position embeddings, causal self-attention, feed-forward layers, residual connections, normalization, logits, cross-entropy, autodiff, AdamW, batching, checkpointing, evaluation, and sampling.

## Override model

Advanced users should replace components without rewriting the framework:

\`\`\`deka
let model = ai.GPT.create(
  config,
  attention: MySparseAttention {},
  optimizer: ai.AdamW.default()
)
\`\`\`

Component interfaces should be explicit and typed. Defaults should be stable library implementations, not compiler magic.

## Required package surface

- \`ai.Embedding\`, \`ai.Linear\`, \`ai.Attention\`
- \`ai.LayerNorm\`, \`ai.TransformerBlock\`, \`ai.GPT\`
- \`ai.Trainer\`, \`ai.AdamW\`, and \`ai.Sampler\`
- tokenizer and checkpoint APIs

## Acceptance criteria

- A complete small GPT can be trained and sampled from in a short Deka program.
- The example does not define tensor storage, matrix multiplication, gradients, or optimizer internals.
- An expert can replace attention, normalization, optimizer, or sampler through documented interfaces.
- Native execution uses the same explicit data/device semantics as lower layers.
- Tests prove causal masking, shifted targets, generation, checkpoint round trips, and deterministic seeded sampling. -codex
